### DISCUSSION TITLE
Fill tool usage / limitations?

### DISCUSSION THREAD

**Discussion by jonnyroth** [gdsfactory/gdsfactory#2648] [Q&A] [answered]
I want to do a fill where the fill geometry gets as close as possible to the edge of a polygon without violating a DRC rule.  Trying to get familiar with the fill tool, I ran the example from the documentation but reduced the margin to zero.  Now the fill overlaps the avoid_layers in some locations.  Is there some trade-off of speed for accuracy in the libraries being used, and am I better off just coding a new fill function for the requirement I'm describing? 
Thanks.

```python

    # modified from example at:
    # https://gdsfactory.github.io/gdsfactory/notebooks/04_components_geometry.html#dummy-fill-tiling
    import gdsfactory as gf
    coupler_lengths = [10, 20, 30, 40, 50, 60, 70, 80]
    coupler_gaps = [0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8]
    delta_lengths = [10, 100, 200, 300, 400, 500, 500]
    mzi = gf.components.mzi_lattice(
        coupler_lengths=coupler_lengths,
        coupler_gaps=coupler_gaps,
        delta_lengths=delta_lengths,
    )

    # Add fill
    c = gf.Component("component_with_fill")
    avoid_layers = [(1, 0)]
    fill_layers = [(2, 0)]
    fill_size = [0.5, 0.5]

    c << gf.fill_rectangle(
        mzi,
        fill_size=fill_size,
        fill_layers=fill_layers,
        margin=0,
        fill_densities=[0.8] * len(fill_layers),
        avoid_layers=avoid_layers,
    )

    c << mzi
    c.show()
```

Here is a zoom-in at the bottom arm of the leftmost mzi:
![fill_example](https://github.com/gdsfactory/gdsfactory/assets/24799153/ad57da70-0f3a-44a1-818a-2a14dc2fa6a0)

**Comment by tvt173** [comment]
hi @jonnyroth , as far as i know, that function is deprecated. you should try the example in gplugins, which uses a klayout backend
https://gdsfactory.github.io/gplugins/notebooks/klayout_dataprep.html#fill

**Reply by tvt173** [reply]
@sebastian-goeldi @joamatab in case i've misspoken...

**Reply by jonnyroth** [reply]
Thanks, @tvt173, I solved my immediate need by writing something custom that fills as part of the component layout flow.  The KLayout integration for GDS postprocessing also looks useful.  I'll keep this open till we get confirmation fill_rectangle is deprecated.

**Reply by joamatab** [reply, marked as answer]
Yes, we recommend using klayout fill with the tile processor

https://gdsfactory.github.io/gdsfactory/notebooks/04_components_geometry.html#dummy-fill-tiling
